presto-spark-base/src/main/java/com/facebook/presto/spark/execution/HttpRemoteTaskResultFetcher.java

```diff
@@ -0,0 +1,168 @@
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.TimeoutException;
+
+import static com.facebook.presto.spi.StandardErrorCode.GENERIC_INTERNAL_ERROR;
+import static com.facebook.presto.spi.StandardErrorCode.SERIALIZED_PAGE_CHECKSUM_ERROR;
+import static com.facebook.presto.spi.page.PagesSerdeUtil.isChecksumValid;
+import static java.lang.String.format;
+import static java.util.Objects.requireNonNull;
+
+public class HttpRemoteTaskResultFetcher
+{
+    private static final Duration FETCH_INTERVAL = new Duration(200,
+            TimeUnit.MILLISECONDS);
+
+    private final ScheduledExecutorService scheduler;
+    private final PrestoSparkHttpWorkerClient workerClient;
+    // Timeout for each fetching request
+    private final Duration requestTimeout;
+    private ScheduledFuture<?> schedulerFuture;
+    private boolean started;
+
+    public HttpRemoteTaskResultFetcher(
+            ScheduledExecutorService scheduler,
+            PrestoSparkHttpWorkerClient workerClient,
+            Duration requestTimeout)
+    {
+        this.scheduler = requireNonNull(scheduler, "scheduler is null");
+        this.workerClient = requireNonNull(workerClient,
+                "workerClient is null");
+        this.requestTimeout = requireNonNull(requestTimeout,
+                "requestTimeout is null");
+    }
+
+    public CompletableFuture<List<SerializedPage>> start(TaskId taskId)
+    {
+        if (started) {
+            throw new PrestoException(GENERIC_INTERNAL_ERROR,
+                    "trying to start an already started TaskResultFetcher for '" + taskId + "'");
+        }
+        CompletableFuture<List<SerializedPage>> future = new CompletableFuture<>();
+        schedulerFuture = scheduler.scheduleAtFixedRate(
+                new HttpRemoteTaskResultFetcherRunner(workerClient, taskId,
+                        future, requestTimeout),
+                0,
+                (long) FETCH_INTERVAL.getValue(),
+                FETCH_INTERVAL.getUnit());
+        started = true;
+        return future.handle(
+                (List<SerializedPage> pages, Throwable throwable) -> {
+                    schedulerFuture.cancel(false);
+                    if (throwable != null) {
+                        throw new CompletionException(throwable.getCause());
+                    }
+                    return pages;
+                });
+    }
+
+    private static class HttpRemoteTaskResultFetcherRunner
+            implements Runnable
+    {
+        private static final DataSize MAX_RESPONSE_SIZE = new DataSize(32,
+                DataSize.Unit.MEGABYTE);
+        private static final int MAX_HTTP_TIMEOUT_RETRIES = 3;
+
+        private final Duration requestTimeout;
+        private int timeoutRetries;
+        private TaskId taskId;
+        private PrestoSparkHttpWorkerClient client;
+        private long token;
+        private CompletableFuture<List<SerializedPage>> future;
+        private List<SerializedPage> serializedPages;
```

Comment:
so of these variables could be final and some are not. For example, `taskId` could be final. Make those final and separate the final vs non-final variables